File: python/feedo/router.py
import httpx
import asyncio
from typing import List, Optional, Dict

DEFAULT_SEEDS = {
    "search": ["http://localhost:8000"],
    "consensus": ["http://localhost:8080"],
    "storage": ["http://localhost:8081"]
}

class NodeRouter:
    def __init__(self, search_seeds: Optional[List[str]] = None, consensus_seeds: Optional[List[str]] = None, storage_seeds: Optional[List[str]] = None):
        self.search_nodes = search_seeds or DEFAULT_SEEDS["search"]
        self.consensus_nodes = consensus_seeds or DEFAULT_SEEDS["consensus"]
        self.storage_nodes = storage_seeds or DEFAULT_SEEDS["storage"]

        self._active_search_node = None
        self._active_consensus_node = None
        self._active_storage_node = None

    async def _find_fastest_node(self, nodes: List[str], health_endpoint: str) -> str:
        async def ping(node: str) -> str:
            async with httpx.AsyncClient() as client:
                url = f"{node}{health_endpoint}"
                response = await client.get(url, timeout=3.0)
                response.raise_for_status()
                return node

        tasks = [asyncio.create_task(ping(node)) for node in nodes]
        
        while tasks:
            done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
            for task in done:
                try:
                    return task.result()
                except Exception:
                    pass
            tasks = list(pending)
            
        print(f"Warning: All seed nodes failed. Falling back to {nodes[0]}")
        return nodes[0]

    async def get_search_node(self) -> str:
        if not self._active_search_node:
            self._active_search_node = await self._find_fastest_node(self.search_nodes, "/explorer/stats")
        return self._active_search_node

    async def get_consensus_node(self) -> str:
        if not self._active_consensus_node:
            self._active_consensus_node = await self._find_fastest_node(self.consensus_nodes, "/grants")
        return self._active_consensus_node

    async def get_storage_node(self) -> str:
        if not self._active_storage_node:
            self._active_storage_node = await self._find_fastest_node(self.storage_nodes, "/api/files/recent")
        return self._active_storage_node

    def invalidate_search_node(self):
        self._active_search_node = None

    def invalidate_consensus_node(self):
        self._active_consensus_node = None

    def invalidate_storage_node(self):
        self._active_storage_node = None

File: typescript/src/router.ts
import axios from 'axios';

export interface FeedoNetworkConfig {
    searchSeeds?: string[];
    consensusSeeds?: string[];
    storageSeeds?: string[];
}

const DEFAULT_SEEDS = {
    // For local dev, we include localhost ports as well as the mainnet/testnet URLs
    search: ["http://localhost:8000"], 
    consensus: ["http://localhost:8080"], // Standard Axum port
    storage: ["http://localhost:8081"]
};

export class NodeRouter {
    private searchNodes: string[];
    private consensusNodes: string[];
    private storageNodes: string[];

    private activeSearchNode: string | null = null;
    private activeConsensusNode: string | null = null;
    private activeStorageNode: string | null = null;

    constructor(config?: FeedoNetworkConfig) {
        this.searchNodes = config?.searchSeeds || DEFAULT_SEEDS.search;
        this.consensusNodes = config?.consensusSeeds || DEFAULT_SEEDS.consensus;
        this.storageNodes = config?.storageSeeds || DEFAULT_SEEDS.storage;
    }

    private async findFastestNode(nodes: string[], healthEndpoint: string): Promise<string> {
        // Ping all nodes and return the first one that resolves successfully
        const promises = nodes.map(async (node) => {
            const url = `${node}${healthEndpoint}`;
            try {
                await axios.get(url, { timeout: 3000 });
                return node;
            } catch (error) {
                throw new Error(`Node ${node} failed ping`);
            }
        });

        try {
            return await Promise.any(promises);
        } catch (error) {
            console.warn(`All seed nodes failed. Falling back to the first node in the list: ${nodes[0]}`);
            return nodes[0];
        }
    }

    public async getSearchNode(): Promise<string> {
        if (!this.activeSearchNode) {
            this.activeSearchNode = await this.findFastestNode(this.searchNodes, '/explorer/stats');
        }
        return this.activeSearchNode;
    }

    public async getConsensusNode(): Promise<string> {
        if (!this.activeConsensusNode) {
            this.activeConsensusNode = await this.findFastestNode(this.consensusNodes, '/grants');
        }
        return this.activeConsensusNode;
    }

    public async getStorageNode(): Promise<string> {
        if (!this.activeStorageNode) {
            this.activeStorageNode = await this.findFastestNode(this.storageNodes, '/api/files/recent');
        }
        return this.activeStorageNode;
    }

    public invalidateSearchNode() {
        this.activeSearchNode = null;
    }

    public invalidateConsensusNode() {
        this.activeConsensusNode = null;
    }

    public invalidateStorageNode() {
        this.activeStorageNode = null;
    }
}
